### PR TITLE
eth/tracers: check the parent callframe's length before removing precompile call

### PR DESCRIPTION
**Pull request by minh-bq**
When using flat call tracer, by default, we remove the call to precompile. However, we should check the length of parent's callframe before doing that because when OnlyTopCall option is provided the call tracer's CaptureExit does not push the child's callframe to parent's callframe. As a result, the parent's callframe is empty, trying to remove the precompile call leads to index out of range. This commit checks the length of parent's callframe before removing precompile call.

### CODE REVIEW

**Comment by s1na** [comment]
Good find!

I'm a bit torn on this. I feel like we should clear the config for the flat call tracer. For example `WithLogs` is also not implemented yet. For that we'll have to change the format and add in logs.

Do you need OnlyTopCall for the flatCallTracer? or discovered this by accident?

**Comment by minh-bq** [comment]
Yes, I also feel that call tracer configs are not intended to be used when flat call tracer is used. Our team does not use the OnlyTopCall for flatCallTracer, just find it during testing.

So, do you prefer this solution, not pass the config from flat call tracer down to inner call tracer?
```
diff --git a/eth/tracers/native/call_flat.go b/eth/tracers/native/call_flat.go
index a9c056f27..1334b8baa 100644
--- a/eth/tracers/native/call_flat.go
+++ b/eth/tracers/native/call_flat.go
@@ -129,6 +129,7 @@ func newFlatCallTracer(ctx *tracers.Context, cfg json.RawMessage) (tracers.Trace
                }
        }
 
+       cfg, _ = json.Marshal(callTracerConfig{})
        tracer, err := tracers.DefaultDirectory.New("callTracer", ctx, cfg)
        if err != nil {
                return nil, err
```

**Comment by s1na** [comment]
Honestly I'd prefer to clear the config for now and only allow those options if someone asks for it. The reason we added the `flatCallTracer` is to have compatibility with other clients. I'd say if we want to e.g. add logs we should consult with them and do it together.

@minh-bq that patch looks good!

**Comment by ziogaschr** [comment]
@s1na yes that a valid point. It's wiser to do this in discussion with other clients when needed.

File: eth/tracers/native/call_flat.go [hunk 1]
```diff
@@ -185,6 +185,14 @@ func (t *flatCallTracer) CaptureExit(output []byte, gasUsed uint64, err error) {
 	if t.config.IncludePrecompiles {
 		return
 	}
+
+	// When OnlyTopCall is provided, the call tracer CaptureExit does
+	// not push the child's callframe to parent's callframe so we don't
+	// need to remove the call to precompile.
+	if len(t.tracer.callstack[len(t.tracer.callstack)-1].Calls) == 0 {
```
[the remark this answers is not in the document]
Comment (minh-bq) [reply eth/tracers/native/call_flat.go:192]: This suggestion looks good to me

**Comment by minh-bq** [comment]
Look like we prefer to clear the config passed to inner call tracer, don't want to edit this PR description and title for reference, I create another PR: https://github.com/ethereum/go-ethereum/pull/27304 for that implementation.

**Pull request closed without merging**